rfcs/20200619-dropout/README.md

```diff
@@ -0,0 +1,425 @@
+such implementation is in progress.
+
+### Random number generation - vectorized implementation
+
+VSL implements several basic RNG engines including VSL_BRNG_MCG31, which is used
+by default in mxnet's dropout (see:
+https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/dropout-inl.h line 108).
+The engine mentioned performs computation of the following recursive formula
+(see: https://en.wikipedia.org/wiki/Linear_congruential_generator):
+
+```python
+def lcg(modulus, a, c, seed):
+    """Linear congruential generator."""
+    while True:
+        seed = (a * seed + c) % modulus
+        yield seed
+```
+
+In VSL MCG31 engine is implemented slightly different - it skips "c" parameter
+(consulted with MKL team). With this change, the core of the formula simplifies
+to:
+
+```python
+seed = (a * seed) % modulus
+```
+
+It is worth noting that in such case:
+
+```python
+seed[1] = seed[0] * A % modulus
+seed[2] = seed[0] * (A * A % modulus) % modulus = seed[1] * A % modulus
+```
+
+So, first N random values can be computed as a multiplication of vector of
+initial seeds (scalar seed broadcasted to vector) by vector of consecutive powers
+of "a" modulo "modulus".
+The latter vectors of random values, as they depend on the former ones, can be
+computed by vector multiplication of previously generated output seed and a
+broadcasted power of a modulo "modulus".
+
+With that said, we can introduce a following vectorized formula for computing
+vectors of random values (with utilization of SIMD instructions):
+
+```
+N - lenght of AVX vector
+
+// first iteration (i=0)
+seed_t[0:N] = [seed, seed, ..., seed]
+a_t[0:N] = [1, A % modulus, A^2 % modulus, ..., A^N % modulus]
+out[0:N] = seed[0:N] * a_t[0:N] % modulus
+
+//following iterations
+aa_t[0:N] = [A^N % modulus, A^N % modulus, ..., A^N % modulus]
+for(i=1:...)
+    out[N*i:N*(i+1)] = out[N*(i-1):N*i] * aa_t[0:N] % modulus
+```
+
+According to GPU team opinion (analysis made by Chereshnev, Eugene) the above
+algorithm should be fairly easily portable to GPU.
+
+### RNG State
+
+Should we store RNG state somewhere ? oneDNN primitives should be stateless.
+
+Possible use case:
+```cpp
+primitve rng(...);
+rng.execute(...) ;
+rng.execute(...) // second exeucte should start from end state of previous execute;
+```
+Let's consider https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/dropout-inl.h
+since line 90. In each time they are creating new stream vsl stream and use
+viRngBernoulli. They don't save stream and viRngBernoulli doesn't save any state.
+The only thing that they're doing is gnenerating new seed in each param.
+
+Possible solution:
+- User should pass seed in execute instead primitive_descriptor (in intial draft,
+seed was passed in the primitive descriptor)
+- User passes seed in execute using DNNL_ARG_SEED. Primitivie in
+DNNL_ARG_SEED memory returns end-state of rng engine. In next calls of 'execute'
```

Comment:
Based on statement above user is allowed to pass new seed. There are two use cases:
- user wants to run dropout with completely new seed - by specifying DNNL_ARG_SEED user can pass to execute in each run seed whatever he wants;
- user in next execution wants to use rng starting from end-state of previous execution - we allowing to do that by returning end seed.
At the end user all time is deciding what seed is passed to execution part.

Reply:
I am not sure about the second use case since it will introduce a primitive state which can cause a few problems, especially with the use of the primitive cache. I would rather force the user to pass a seed, and default to some fixed value if the seed is not provided.

Reply:
1. In our proposition user always must pass DNNL_ARG_SEED.
2. Primitive is not storing in it's memory any state between executions. Let's consider proposed MCG31m1. Starting seed (base on passed initial seed) for each thread is computed as described in document MCG31m1 vectorised algorithm. We want also to compute "end seed" and return it to user the same way as we're returning tensors. User is deciding what seed will pass for next execution. He can take returned value and pass as DNNL_ARG_SEED or generate new custom value. Why do you think that it will affect primitive cache ?